### PR TITLE
Fix deadlocks in pytest-cov runs

### PR DESCRIPTION
**Pull request by cdiener**
* [ ] fixes deadlocks in the CI
* [X] description of feature/fix
* [X] tests added/passed
* [ ] add an entry to the [next release](../release-notes/next-release.md)

This combines the changes in #1188 with main guards for all the parallel tests which will avoid deadlocks in pytest-cov.

### CODE REVIEW

File: src/cobra/sampling/optgp.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 import numpy as np
 import pandas as pd
+from pydantic import NoneIsAllowedError
```
Comment (synchon) [inline src/cobra/sampling/optgp.py]: Not sure where this import is being used.

**Review by synchon — COMMENTED** [review]
Just a minor comment about the import. Looks great and the CI works perfectly. Thanks!

**Comment by Midnighter** [comment]
I had one doubt when playing around with this myself. Are you sure those tests are still properly run? Or do they pass because they are essentially skipped due to the main guard?

**Comment by cdiener** [comment]
I mean they are actually passing, so I would assume they do. I am not 100% certain if they get picked up by coverage.

**Comment by Midnighter** [comment]
Well,

```py
def test_foo():
    pass
```

is a valid passing test case. So tests where the `__name__` is not `"__main__"` would also simply pass.

**Comment by Midnighter** [comment]
I just ran tests on devel and for this branch, coverage looks the same and timings are equivalent. So I guess it works correctly.

**Comment by Midnighter** [comment]
When you address @synchon 's comment, I think it's good to go in.

**Comment by Midnighter** [comment]
Actually, just to be sure I raised an exception within one of those if blocks in the tests and pytest skips right over it.

**Comment by cdiener** [comment]
Yeah, you're right! This is so weird because it is exactly what pytest-cov says one should do in their own docs: https://pytest-cov.readthedocs.io/en/latest/subprocess-support.html .

**Comment by cdiener** [comment]
That makes it difficult since it is definitely a deadlock, at least when running pytest-cov. I did see some other projects have the same issue, and some of that seem to stem from [open bugs in Python itself](https://github.com/pytest-dev/pytest-cov/issues/492). The only other alternative I see is trying to separate tests and coverage and skip multiprocessing tests in coverage only...

**Comment by Midnighter** [comment]
Hmm, should we just add `# pragma: no cover` to those test cases that use the `ProcessPool`?

**Comment by cdiener** [comment]
The question is if it would still run it in the coverage process and exclude it afterwards, or not run it at all. There is also a chance that pytest itself might deadlock, even though I think that is unlikely.

**Comment by cdiener** [comment]
Though I would like some workaround for now as this breaks our releaser. I also tested `pragma: no cover` and `pytest.mark.no_cover` and both will still run the code (just not record coverage) so it wouldn't fix our issue. We could do it with a `TEST_MP` environment variable and `pytest.mark.skipif` directive...

**Comment by cdiener** [comment]
Here is an implementation based on the previous command. It seems to work.

**Comment by synchon** [comment]
Looks cleaner and works. 🎉

**Review by synchon — APPROVED** [review]
Thanks for the effort! 🙏🏻

**Review by Midnighter — APPROVED** [review]
Feels like a pretty convoluted solution but I don't have a better idea. Thanks for working this out 👍🏼

**Comment by cdiener** [comment]
Yeah, fixing the source issue would have been better. Maybe it will fix itself in the future...

**Pull request merged**